File: python/carrier_verifier.py
"""
Carrier Verification Module
Implements the FMCSA-Based Decision Framework for carrier onboarding
"""

import json
import requests
from datetime import datetime
from dateutil import parser
from typing import Dict, List, Tuple, Optional


class CarrierVerifier:
    """Main class for carrier verification using FMCSA/SAFER data"""
    
    def __init__(self, api_key: str, api_base_url: str = "https://saferwebapi.com/v2/usdot/snapshot"):
        self.api_key = api_key
        self.api_base_url = api_base_url
        self.headers = {"x-api-key": api_key}
    
    def fetch_carrier_data(self, usdot: str) -> Dict:
        """Fetch carrier snapshot data from SaferWebAPI"""
        url = f"{self.api_base_url}/{usdot}"
        try:
            response = requests.get(url, headers=self.headers)
            response.raise_for_status()
            return response.json()
        except requests.exceptions.HTTPError as e:
            if response.status_code == 401:
                raise Exception(f"Unauthorized (401): Invalid or expired API key. Please check your API key in .env file.")
            else:
                raise Exception(f"HTTP Error {response.status_code}: {str(e)}")
        except requests.exceptions.RequestException as e:
            raise Exception(f"Failed to fetch carrier data: {str(e)}")
    
    def verify_authority_status(self, data: Dict) -> str:
        """Rule 2: Authority Status"""
        operating_status = data.get("operating_status", "").upper()
        
        if operating_status in ["ACTIVE", "AUTHORIZED"]:
            return "ACCEPT"
        elif operating_status in ["NOT AUTHORIZED", "INACTIVE", "OUT OF SERVICE"]:
            return "DENY"
        else:
            return "REVIEW"
    
    def verify_data_freshness(self, data: Dict) -> str:
        """Rule 3: Data Freshness (MCS-150 Update)"""
        mcs_150_date_str = data.get("mcs_150_form_date")
        if not mcs_150_date_str:
            return "DENY"
        
        try:
            mcs_150_date = parser.parse(mcs_150_date_str)
            months_old = (datetime.now() - mcs_150_date).days / 30.0
            
            if months_old <= 12:
                return "ACCEPT"
            elif months_old <= 24:
                return "REVIEW"
            else:
                return "DENY"
        except Exception:
            return "DENY"
    
    def verify_vehicle_oos_rate(self, data: Dict) -> str:
        """Rule 4: Vehicle Out-of-Service Rate"""
        us_inspections = data.get("us_inspections", {})
        vehicle = us_inspections.get("vehicle", {})
        
        inspections = int(vehicle.get("inspections", 0))
        out_of_service = int(vehicle.get("out_of_service", 0))
        
        if inspections == 0:
            return "REVIEW"  # No data
        
        oos_rate = (out_of_service / inspections) * 100
        
        if oos_rate <= 25:
            return "ACCEPT"
        elif oos_rate <= 30:
            return "REVIEW"
        else:
            return "DENY"
    
    def verify_driver_oos_rate(self, data: Dict) -> str:
        """Rule 5: Driver Out-of-Service Rate"""
        us_inspections = data.get("us_inspections", {})
        driver = us_inspections.get("driver", {})
        
        inspections = int(driver.get("inspections", 0))
        out_of_service = int(driver.get("out_of_service", 0))
        
        if inspections == 0:
            return "REVIEW"  # No data
        
        oos_rate = (out_of_service / inspections) * 100
        
        if oos_rate <= 7:
            return "ACCEPT"
        elif oos_rate <= 10:
            return "REVIEW"
        else:
            return "DENY"
    
    def verify_crash_history(self, data: Dict) -> str:
        """Rule 6: Crash History"""
        us_crashes = data.get("united_states_crashes", {})
        
        tow = int(us_crashes.get("tow", 0))
        injury = int(us_crashes.get("injury", 0))
        fatal = int(us_crashes.get("fatal", 0))
        
        if fatal > 0 or injury >= 2:
            return "DENY"
        elif injury == 1:
            return "REVIEW"
        elif tow <= 1:
            return "ACCEPT"
        else:
            return "REVIEW"
    
    def verify_inspection_volume(self, data: Dict) -> str:
        """Rule 7: Inspection Volume"""
        us_inspections = data.get("us_inspections", {})
        
        vehicle_inspections = int(us_inspections.get("vehicle", {}).get("inspections", 0))
        driver_inspections = int(us_inspections.get("driver", {}).get("inspections", 0))
        hazmat_inspections = int(us_inspections.get("hazmat", {}).get("inspections", 0))
        
        total_inspections = vehicle_inspections + driver_inspections + hazmat_inspections
        
        if total_inspections < 10:
            return "REVIEW"
        else:
            return "ACCEPT"
    
    def verify_fleet_size(self, data: Dict) -> str:
        """Rule 8: Fleet Size"""
        power_units = int(data.get("power_units", 0))
        
        if power_units >= 5:
            return "ACCEPT"
        elif power_units >= 2:
            return "REVIEW"
        else:
            return "DENY"
    
    def verify_safety_rating(self, data: Dict) -> str:
        """Rule 10: Safety Rating"""
        safety_rating = data.get("safety_rating", "").upper() if data.get("safety_rating") else "NOT RATED"
        
        if safety_rating == "SATISFACTORY":
            return "ACCEPT"
        elif safety_rating in ["CONDITIONAL", "UNSATISFACTORY"]:
            return "DENY"
        else:  # NOT RATED or None
            return "REVIEW"
    
    def verify_carrier(self, usdot: str, phone_number: Optional[str] = None) -> Dict:
        """
        Main verification method
        
        Args:
            usdot: USDOT number to verify
            phone_number: Optional phone number for identity verification (Rule 1)
        
        Returns:
            Dictionary with verification results and final decision
        """
        # Fetch carrier data
        data = self.fetch_carrier_data(usdot)
        
        results = {
            "usdot": usdot,
            "carrier_name": data.get("legal_name", "Unknown"),
            "dba_name": data.get("dba_name", ""),
            "checks": {},
            "final_decision": None,
            "reason": ""
        }
        
        # Rule 1: Identity & Ownership Verification (if phone provided)
        if phone_number:
            fmcsa_phone = data.get("phone", "").replace(" ", "").replace("-", "").replace("(", "").replace(")", "")
            provided_phone = phone_number.replace(" ", "").replace("-", "").replace("(", "").replace(")", "")
            if fmcsa_phone == provided_phone:
                results["checks"]["identity_verification"] = "PASS"
            else:
                results["checks"]["identity_verification"] = "FAIL"
                results["final_decision"] = "AUTO-REJECT"
                results["reason"] = "Phone number mismatch"
                return results
        else:
            results["checks"]["identity_verification"] = "SKIPPED"
        
        # Rule 2: Authority Status
        results["checks"]["authority_status"] = self.verify_authority_status(data)
        
        # Rule 3: Data Freshness
        results["checks"]["data_freshness"] = self.verify_data_freshness(data)
        
        # Rule 4: Vehicle OOS Rate
        results["checks"]["vehicle_oos_rate"] = self.verify_vehicle_oos_rate(data)
        
        # Rule 5: Driver OOS Rate
        results["checks"]["driver_oos_rate"] = self.verify_driver_oos_rate(data)
        
        # Rule 6: Crash History
        results["checks"]["crash_history"] = self.verify_crash_history(data)
        
        # Rule 7: Inspection Volume
        results["checks"]["inspection_volume"] = self.verify_inspection_volume(data)
        
        # Rule 8: Fleet Size
        results["checks"]["fleet_size"] = self.verify_fleet_size(data)
        
        # Rule 10: Safety Rating
        results["checks"]["safety_rating"] = self.verify_safety_rating(data)
        
        # Final Decision Logic
        check_results = list(results["checks"].values())
        deny_count = check_results.count("DENY")
        review_count = check_results.count("REVIEW")
        
        if deny_count > 0:
            results["final_decision"] = "AUTO-REJECT"
            results["reason"] = f"{deny_count} DENY flag(s) found"
        elif review_count >= 2:
            results["final_decision"] = "CONDITIONAL_APPROVAL"
            results["reason"] = f"{review_count} REVIEW flag(s) - requires increased monitoring"
        else:
            results["final_decision"] = "FULL_APPROVAL"
            results["reason"] = "All checks passed"
        
        results["raw_data"] = data
        return results

File: csharp/CarrierVerifier.cs
using RestSharp;
using Newtonsoft.Json;
using System.Text.RegularExpressions;

namespace CarrierVerification;

public class CarrierVerifier
{
    private readonly string _apiKey;
    private readonly string _apiBaseUrl;

    public CarrierVerifier(string apiKey, string apiBaseUrl = "https://saferwebapi.com/v2/usdot/snapshot")
    {
        _apiKey = apiKey;
        _apiBaseUrl = apiBaseUrl;
    }

    public CarrierData FetchCarrierData(string usdot)
    {
        var client = new RestClient(_apiBaseUrl);
        var request = new RestRequest($"/{usdot}", Method.Get);
        request.AddHeader("x-api-key", _apiKey);

        var response = client.Execute(request);
        
        if (!response.IsSuccessful)
        {
            throw new Exception($"Failed to fetch carrier data: {response.ErrorMessage ?? response.StatusCode.ToString()}");
        }

        return JsonConvert.DeserializeObject<CarrierData>(response.Content!) 
            ?? throw new Exception("Failed to deserialize carrier data");
    }

    public string VerifyAuthorityStatus(CarrierData data)
    {
        var operatingStatus = data.OperatingStatus?.ToUpper() ?? "";

        if (operatingStatus.Contains("ACTIVE") || operatingStatus.Contains("AUTHORIZED"))
        {
            return "ACCEPT";
        }
        else if (operatingStatus.Contains("NOT AUTHORIZED") || 
                 operatingStatus.Contains("INACTIVE") || 
                 operatingStatus.Contains("OUT OF SERVICE"))
        {
            return "DENY";
        }
        else
        {
            return "REVIEW";
        }
    }

    public string VerifyDataFreshness(CarrierData data)
    {
        if (string.IsNullOrEmpty(data.Mcs150FormDate))
        {
            return "DENY";
        }

        try
        {
            if (DateTime.TryParse(data.Mcs150FormDate, out var mcs150Date))
            {
                var monthsOld = (DateTime.Now - mcs150Date).TotalDays / 30.0;

                if (monthsOld <= 12)
                {
                    return "ACCEPT";
                }
                else if (monthsOld <= 24)
                {
                    return "REVIEW";
                }
                else
                {
                    return "DENY";
                }
            }
        }
        catch
        {
            // Fall through to DENY
        }

        return "DENY";
    }

    public string VerifyVehicleOosRate(CarrierData data)
    {
        var vehicle = data.UsInspections?.Vehicle;
        if (vehicle == null)
        {
            return "REVIEW";
        }

        var inspections = int.TryParse(vehicle.Inspections, out var insp) ? insp : 0;
        var outOfService = int.TryParse(vehicle.OutOfService, out var oos) ? oos : 0;

        if (inspections == 0)
        {
            return "REVIEW";
        }

        var oosRate = (outOfService / (double)inspections) * 100;

        if (oosRate <= 25)
        {
            return "ACCEPT";
        }
        else if (oosRate <= 30)
        {
            return "REVIEW";
        }
        else
        {
            return "DENY";
        }
    }

    public string VerifyDriverOosRate(CarrierData data)
    {
        var driver = data.UsInspections?.Driver;
        if (driver == null)
        {
            return "REVIEW";
        }

        var inspections = int.TryParse(driver.Inspections, out var insp) ? insp : 0;
        var outOfService = int.TryParse(driver.OutOfService, out var oos) ? oos : 0;

        if (inspections == 0)
        {
            return "REVIEW";
        }

        var oosRate = (outOfService / (double)inspections) * 100;

        if (oosRate <= 7)
        {
            return "ACCEPT";
        }
        else if (oosRate <= 10)
        {
            return "REVIEW";
        }
        else
        {
            return "DENY";
        }
    }

    public string VerifyCrashHistory(CarrierData data)
    {
        var crashes = data.UnitedStatesCrashes;
        if (crashes == null)
        {
            return "ACCEPT";
        }

        var fatal = crashes.Fatal;
        var injury = crashes.Injury;
        var tow = crashes.Tow;

        if (fatal > 0 || injury >= 2)
        {
            return "DENY";
        }
        else if (injury == 1)
        {
            return "REVIEW";
        }
        else if (tow <= 1)
        {
            return "ACCEPT";
        }
        else
        {
            return "REVIEW";
        }
    }

    public string VerifyInspectionVolume(CarrierData data)
    {
        var inspections = data.UsInspections;
        if (inspections == null)
        {
            return "REVIEW";
        }

        var vehicleInspections = int.TryParse(inspections.Vehicle?.Inspections, out var v) ? v : 0;
        var driverInspections = int.TryParse(inspections.Driver?.Inspections, out var d) ? d : 0;
        var hazmatInspections = int.TryParse(inspections.Hazmat?.Inspections, out var h) ? h : 0;

        var totalInspections = vehicleInspections + driverInspections + hazmatInspections;

        if (totalInspections < 10)
        {
            return "REVIEW";
        }
        else
        {
            return "ACCEPT";
        }
    }

    public string VerifyFleetSize(CarrierData data)
    {
        var powerUnits = data.PowerUnits;

        if (powerUnits >= 5)
        {
            return "ACCEPT";
        }
        else if (powerUnits >= 2)
        {
            return "REVIEW";
        }
        else
        {
            return "DENY";
        }
    }

    public string VerifySafetyRating(CarrierData data)
    {
        var safetyRating = data.SafetyRating?.ToUpper() ?? "NOT RATED";

        if (safetyRating == "SATISFACTORY")
        {
            return "ACCEPT";
        }
        else if (safetyRating == "CONDITIONAL" || safetyRating == "UNSATISFACTORY")
        {
            return "DENY";
        }
        else
        {
            return "REVIEW";
        }
    }

    public VerificationResult VerifyCarrier(string usdot, string? phoneNumber = null)
    {
        var data = FetchCarrierData(usdot);
        var result = new VerificationResult
        {
            Usdot = usdot,
            CarrierName = data.LegalName ?? "Unknown",
            DbaName = data.DbaName ?? "",
            RawData = data
        };

        // Rule 1: Identity & Ownership Verification
        if (!string.IsNullOrEmpty(phoneNumber))
        {
            var fmcsaPhone = NormalizePhone(data.Phone ?? "");
            var providedPhone = NormalizePhone(phoneNumber);
            
            if (fmcsaPhone == providedPhone)
            {
                result.Checks["identity_verification"] = "PASS";
            }
            else
            {
                result.Checks["identity_verification"] = "FAIL";
                result.FinalDecision = "AUTO-REJECT";
                result.Reason = "Phone number mismatch";
                return result;
            }
        }
        else
        {
            result.Checks["identity_verification"] = "SKIPPED";
        }

        // Rule 2: Authority Status
        result.Checks["authority_status"] = VerifyAuthorityStatus(data);

        // Rule 3: Data Freshness
        result.Checks["data_freshness"] = VerifyDataFreshness(data);

        // Rule 4: Vehicle OOS Rate
        result.Checks["vehicle_oos_rate"] = VerifyVehicleOosRate(data);

        // Rule 5: Driver OOS Rate
        result.Checks["driver_oos_rate"] = VerifyDriverOosRate(data);

        // Rule 6: Crash History
        result.Checks["crash_history"] = VerifyCrashHistory(data);

        // Rule 7: Inspection Volume
        result.Checks["inspection_volume"] = VerifyInspectionVolume(data);

        // Rule 8: Fleet Size
        result.Checks["fleet_size"] = VerifyFleetSize(data);

        // Rule 10: Safety Rating
        result.Checks["safety_rating"] = VerifySafetyRating(data);

        // Final Decision Logic
        var denyCount = result.Checks.Values.Count(v => v == "DENY");
        var reviewCount = result.Checks.Values.Count(v => v == "REVIEW");

        if (denyCount > 0)
        {
            result.FinalDecision = "AUTO-REJECT";
            result.Reason = $"{denyCount} DENY flag(s) found";
        }
        else if (reviewCount >= 2)
        {
            result.FinalDecision = "CONDITIONAL_APPROVAL";
            result.Reason = $"{reviewCount} REVIEW flag(s) - requires increased monitoring";
        }
        else
        {
            result.FinalDecision = "FULL_APPROVAL";
            result.Reason = "All checks passed";
        }

        return result;
    }

    private string NormalizePhone(string phone)
    {
        return Regex.Replace(phone, @"[\s\-\(\)]", "");
    }
}

